### PR TITLE
IMU correction quaternion option

### PR DESCRIPTION
**Pull request by ojura**
If one wants to apply the correction quaternion from Cartographer's IMU auto-calibration, this per-trajectory option enables using it verbatim, without having to modify the TF tree.

### CODE REVIEW

File: cartographer_ros_msgs/msg/TrajectoryOptions.msg [hunk 1]
```diff
@@ -29,6 +29,7 @@ float64 odometry_sampling_ratio
 float64 fixed_frame_pose_sampling_ratio
 float64 imu_sampling_ratio
 float64 landmarks_sampling_ratio
+geometry_msgs/Quaternion imu_correction
```
[the remark this answers is not in the document]
Comment (ojura) [reply cartographer_ros_msgs/msg/TrajectoryOptions.msg]: > Yes, I have copy-pasted the quaternion from Cartographer's log output and converted it to a urdf-readable roll-pitch-yaw angle using tf python and I didn't enjoy it.

We're on the same wavelength :smile: I like your idea as well. In that case, we can open an issue or even an rfc for this in the libcartographer repo.

[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer_ros_msgs/msg/TrajectoryOptions.msg]: Sounds good. For additional log output and no new dependencies we don't need an rfc.

**Comment by ojura** [comment]
Replaced by https://github.com/googlecartographer/cartographer/issues/1516

**Comment by Ellon** [comment]
Hi @ojura,

I'm trying to understand the meaning of `imu_calibration` quaternion estimated by Cartographer and outputted in the optimization logs. From what I see on this PR it seems that you understood well its meaning, so would you mind confirming if the following is correct?

1. From [this line](https://github.com/googlecartographer/cartographer_ros/pull/1025/files#diff-6048bec7723ede35a5a7c6b73e2bd2a0R143) in your commits I understood that `imu_calibration` represents the "IMU frame to IMU calibrated frame", in a way that if I apply this rotation to data expressed into the IMU frame I would have the data in the IMU calibrated frame. Is that so?

2. If I want to correct my TF tree by hand with this calibration, could it be done by setting the new IMU frame to be the old IMU frame composed with the inverse of the rotation contained on `imu_calibration`?

Thanks in advance!

**Comment by ojura** [comment]
Hi @Ellon,

you are correct about 1). About 2, it’s always a PITA to determine if it is indeed that transform or its inverse. The way I would verify would be to compare imu_correction * old  sensor_to_tracking * imu_ data vs new sensor_to_tracking (with correction)  * imu_data.

**Comment by RonaldEnsing** [comment]
I agree that it would be good to have the extrinsics configured on a single place. Having the extrinsics both in the URDF file and a lua file can be confusing.

Is there any plan to fix this in cartographer [1]? Master branch moved forward and is conflicting with the branch of this PR in the mean time. I really think this is a very useful feature.

[1] https://github.com/googlecartographer/cartographer/issues/1516

**Pull request closed without merging**